File: 0907-koko-eating-bananas/0907-koko-eating-bananas.js
/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */
var minEatingSpeed = function(piles, h) {
    const maxBan = piles.reduce((a, b) => Math.max(a, b), 0);
    let res = maxBan;
    if (h === piles.length) return res;
    function getTime(n) {
        let time = 0;
        piles.forEach(p => time += Math.ceil(p / n));
        return time;
    }
    let [i, j] = [1, maxBan];
    while (i <= j) {
        const mid = Math.floor((i + j) / 2);
        const time = getTime(mid);
        if (time <= h) {
            res = mid;
            j = mid - 1;
        } else {
            i = mid + 1;
        }
    }
    return res;
};

File: 0907-koko-eating-bananas/0907-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        max_ban = max(piles)
        res = max_ban

        if h == len(piles):
            return max_ban

        def getTime(n):
            time = 0
            for p in piles:
                time = time + ceil(p / n)
            return time

        i, j = 1, max_ban

        while i <= j:
            mid = (i + j) // 2
            time = getTime(mid)
            if time <= h:
                res = mid
                j = mid - 1
            else:
                i = mid + 1
        return res